### PR TITLE
Add type generation for TypeScript

### PR DESCRIPTION
**Pull request by maltejur**
Like discussed in https://github.com/directus/directus/discussions/5418#discussioncomment-1422801, https://github.com/directus/directus/issues/7909#issuecomment-915348857 and https://github.com/directus/directus/discussions/5418#discussioncomment-1021942, this adds the route `/server/specs/ts` that will generate TypeScript type definitions of the current directus project.

Screenshot_20211004_152912

Right now this also generates a new type, `CustomDirectusTypes`, that enables use of the types with the `@directus/sdk`. It looks like this

```ts
export type CustomDirectusTypes = {
  directus_collections: DirectusCollections;
  directus_migrations: DirectusMigrations;
  directus_activity: DirectusActivity;
  directus_folders: DirectusFolders;
  directus_fields: DirectusFields;
// ...
```

And can be used like this

```ts
import { Directus } from "@directus/sdk";
import { CustomDirectusTypes } from "./types";

const directus = new Directus<CustomDirectusTypes>("<directus url>");
```

---
Note: Until this pull request is merged you can use this extension to get the same functionality https://github.com/maltejur/directus-extension-generate-types

### CODE REVIEW

File: api/src/controllers/server.ts [hunk 1]
```diff
@@ -47,6 +47,20 @@ router.get(
...
+router.get(
+	'/specs/ts',
+	asyncHandler(async (req, res) => {
+		const service = new SpecificationService({
+			accountability: req.accountability,
+			schema: req.schema,
+		});
+
+		res.type('text/plain');
+		res.send(await service.typescript.generate());
```
Comment (maltejur) [inline api/src/controllers/server.ts:59]: Should this route prompt a file download like `/specs/graphql`?

If so, this should be added:
```suggestion
		res.attachment('types.ts');
		res.type('text/plain');
		res.send(await service.typescript.generate());
```

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Thanks @maltejur!

File: api/src/services/specifications.ts [hunk 2]
```diff
@@ -579,3 +581,62 @@ class GraphQLSpecsService implements SpecificationSubService {
...
+	}
+
+	getPascalName(snakeName: string) {
+		return snakeName
+			.split('_')
+			.map((x) => x.charAt(0).toUpperCase() + x.slice(1))
+			.join('');
+	}
+
+	getType(directusType: string) {
```
[the remark this answers is not in the document]
Comment (maltejur) [reply api/src/services/specifications.ts:636]: I don't know if I understand you correctly, but the code should already do that in line 614

https://github.com/directus/directus/blob/0d34333346346f29f7f3211f5d907c6bff00f5d7/api/src/services/specifications.ts#L614

File: api/src/controllers/server.ts (hunk 1 above)
Comment (rijkvanzanten) [reply api/src/controllers/server.ts:59]: Yes, lets do that 👍🏻

File: api/src/services/specifications.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/services/specifications.ts:636]: Ohhh my b! Overlooked that earlier 👍🏻  I expected to find that under 'getType'

File: api/src/services/specifications.ts [hunk 3]
```diff
@@ -579,3 +581,62 @@ class GraphQLSpecsService implements SpecificationSubService {
...
+		customDirectusTypes += '\n}';
+		return customDirectusTypes;
+	}
+
+	getPascalName(snakeName: string) {
+		return snakeName
+			.split('_')
+			.map((x) => x.charAt(0).toUpperCase() + x.slice(1))
+			.join('');
+	}
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/services/specifications.ts:634]: > This shouldn't be a problem. UserList would still be called UserList and user_list would become UserList.

Haha that's exactly the problem! Now you have 2 different collections with 2 different types named the same 😬

[the remark this answers is not in the document]
Comment (maltejur) [reply api/src/services/specifications.ts:656]: Okay now if two pascal case names would collide, the original name is used 

Screenshot_20211004_164648

**Comment by rijkvanzanten** [comment]
@maltejur Question! Should we only return custom fields that were added to the system collections? The SDK already contains the proper types for the system handlers, and for other system uses it might be better to rely on the types included in `@directus/shared`, as they more accurately reflect the API IO 🤔 

(Just noticed DirectusCollections, which technically has the wrong signature by auto-generating it like this)

**Comment by maltejur** [comment]
> Maybe I will use your previous commit to extract them

Just in case you actually want to do that, here are the currently generated system types.

<details>

```ts
export type DirectusCollections = {
  collection?: string;
  icon?: string;
  note?: string;
  display_template?: string;
  hidden: boolean;
  singleton: boolean;
  translations?: any[] | Record<string, any>;
  archive_field?: string;
  archive_app_filter: boolean;
  archive_value?: string;
  unarchive_value?: string;
  sort_field?: string;
  accountability?: string;
  color?: string;
  item_duplication_fields?: any[] | Record<string, any>;
  collection_divider: string;
  archive_divider: string;
  sort_divider: string;
  accountability_divider: string;
  duplication_divider: string;
}

export type DirectusMigrations = {
  version: string;
  name: string;
  timestamp?: string;
}

export type DirectusActivity = {
  id: number;
  action: string;
  user?: string;
  timestamp: string;
  ip: string;
  user_agent?: string;
  collection: string;
  item: string;
  comment?: string;
  revisions: string;
}

export type DirectusFolders = {
  id: string;
  name: string;
  parent?: string;
}

export type DirectusRelations = {
  id: number;
  many_collection: string;
  many_field: string;
  one_collection?: string;
  one_field?: string;
  one_collection_field?: string;
  one_allowed_collections?: any[] | Record<string, any>;
  junction_field?: string;
  sort_field?: string;
  one_deselect_action: string;
}

export type DirectusRevisions = {
  id: number;
  activity: number;
  collection: string;
  item: string;
  data?: any[] | Record<string, any>;
  delta?: any[] | Record<string, any>;
  parent?: number;
}

export type DirectusSessions = {
  token: string;
  user: string;
  expires: string;
  ip?: string;
  user_agent?: string;
  data?: any[] | Record<string, any>;
}

export type DirectusDashboards = {
  id?: string;
  name: string;
  icon: string;
  note?: string;
  date_created?: string;
  user_created?: string;
  panels: string;
}

export type DirectusPanels = {
  id?: string;
  dashboard: string;
  name?: string;
  icon?: string;
  color?: string;
  show_header: boolean;
  note?: string;
  type: string;
  position_x: number;
  position_y: number;
  width: number;
  height: number;
  options?: any[] | Record<string, any>;
  date_created?: string;
  user_created?: string;
}

export type DirectusFiles = {
  id: string;
  storage: string;
  filename_disk?: string;
  filename_download: string;
  title?: string;
  type?: string;
  folder?: string;
  uploaded_by?: string;
  uploaded_on: string;
  modified_by?: string;
  modified_on: string;
  charset?: string;
  filesize?: number;
  width?: number;
  height?: number;
  duration?: number;
  embed?: string;
  description?: string;
  location?: string;
  tags?: any[] | Record<string, any>;
  metadata?: any[] | Record<string, any>;
  test?: string;
  storage_divider: string;
}

export type DirectusSettings = {
  id: number;
  project_name: string;
  project_url?: string;
  project_color?: string;
  project_logo?: string;
  public_foreground?: string;
  public_background?: string;
  public_note?: string;
  auth_login_attempts?: number;
  auth_password_policy?: string;
  storage_asset_transform?: string;
  storage_asset_presets?: any[] | Record<string, any>;
  custom_css?: string;
  storage_default_folder?: string;
  basemaps?: any[] | Record<string, any>;
  mapbox_key?: string;
  module_bar?: any[] | Record<string, any>;
  public_divider: string;
  security_divider: string;
  files_divider: string;
  overrides_divider: string;
  map_divider: string;
  modules_divider: string;
}

export type DirectusPermissions = {
  id: number;
  role?: string;
  collection: string;
  action: string;
  permissions?: any[] | Record<string, any>;
  validation?: any[] | Record<string, any>;
  presets?: any[] | Record<string, any>;
  fields?: any[] | Record<string, any>;
}

export type DirectusUsers = {
  id?: string;
  first_name?: string;
  last_name?: string;
  email?: string;
  password?: string;
  location?: string;
  title?: string;
  description?: string;
  tags?: any[] | Record<string, any>;
  avatar?: string;
  language?: string;
  theme?: string;
  tfa_secret?: string;
  status: string;
  role?: string;
  token?: string;
  last_access?: string;
  last_page?: string;
  provider: string;
  external_identifier?: string;
  preferences_divider: string;
  admin_divider: string;
}

export type DirectusRoles = {
  id?: string;
  name: string;
  icon: string;
  description?: string;
  ip_access?: any[] | Record<string, any>;
  enforce_tfa: boolean;
  collection_list?: any[] | Record<string, any>;
  admin_access: boolean;
  app_access: boolean;
  users: string;
}

export type DirectusWebhooks = {
  id: number;
  name: string;
  method: string;
  url: string;
  status: string;
  data: boolean;
  actions: any[] | Record<string, any>;
  collections: any[] | Record<string, any>;
  triggers_divider: string;
}

export type DirectusFields = {
  id: number;
  collection: string;
  field: string;
  special?: any[] | Record<string, any>;
  interface?: string;
  options?: any[] | Record<string, any>;
  display?: string;
  display_options?: any[] | Record<string, any>;
  readonly: boolean;
  hidden: boolean;
  sort?: number;
  width?: string;
  translations?: any[] | Record<string, any>;
  note?: string;
  conditions?: any[] | Record<string, any>;
  required?: boolean;
  group?: string;
}

export type CustomDirectusTypes = {
  directus_collections: DirectusCollections;
  directus_migrations: DirectusMigrations;
  directus_activity: DirectusActivity;
  directus_folders: DirectusFolders;
  directus_relations: DirectusRelations;
  directus_revisions: DirectusRevisions;
  directus_sessions: DirectusSessions;
  directus_dashboards: DirectusDashboards;
  directus_panels: DirectusPanels;
  directus_files: DirectusFiles;
  directus_settings: DirectusSettings;
  directus_permissions: DirectusPermissions;
  directus_users: DirectusUsers;
  directus_roles: DirectusRoles;
  directus_webhooks: DirectusWebhooks;
  directus_fields: DirectusFields;
  directus_presets: DirectusPresets;
}
```

</details>

Some of the indexes are to be nullable which doesn't seem right to me and should probably be changed, but that's how its reported from the api. Also at least the json fields should be replaced with the real type.

**Comment by joselcvarela** [comment]
Thank you so much @maltejur ❤

**Comment by joselcvarela** [comment]
@maltejur I think @nickrum was giving an example.
It is not literally the `custom_directus_types` but, `custom_directus_files`, `custom_directus_users`, etc. So i think your latest commit could not solve that issue. 
Maybe we can keep only with the user defined models.
It can be improved later, I think.

**Comment by maltejur** [comment]
> I do like the idea of PascalCasing all the types to make it more "typescript native", however, it does seem to cause quite a bit of naming conflict potential. Should we simply opt for using the raw collection key instead?

I *hope* there aren't any edge cases left that would cause collisions, and even there were some, they would probably only occur if the user names their collections in a really weird way. So I would still like to have the pascal casing.

**Comment by rijkvanzanten** [comment]
@joselcvarela lets not over-complicate it for this first MVP. We can start with 1 way that we know works, and can add more options in the future if people run into actual use cases 🙂 I just wanna make sure that that first way is the correct way

**Comment by nickrum** [comment]
> Yes it would. Thanks for the catch. Now it will just keep the snake case `custom_directus_types`.

Nice! But, won't there be a conflict if a collection is called `CustomDirectusTypes` then?

It would be nice if Typescript would support _default-exporting_ a type like `export default type = { ... }` to get around this issue....

**Comment by maltejur** [comment]
> Nice! But, won't there be a conflict if a collection is called `CustomDirectusTypes` then?
> 
> It would be nice if Typescript would support _default-exporting_ a type like `export default type = { ... }` to get around this issue....

Yes there would, but you wouldn't be able to call your collection like that in the directus interface. Nonetheless I've now also fixed that and in the case `CustomDirectusTypes` already exists, the custom types will be called `_CustomDirectusTypes`. You *could* still break that if you tried really hard, but I think that should be enough.

**Comment by maltejur** [comment]
@joselcvarela I can try to do that but I am not sure where the types should be put. In [`shared/src/types`](https://github.com/directus/directus/tree/main/packages/shared/src/types), there already are types with some overlap to the types needed in the SDK, like `settings.ts`. But some types are missing and some are just different, like `panels.ts`.

**Comment by joselcvarela** [comment]
Maybe you can create a file`shared/src/types/system.ts` with the generated files.
Yes, it's not good having both multiple types representing the same thing, but I think it's better to have redundant information then no have information at all

**Comment by maltejur** [comment]
@joselcvarela okay, I've added that in this pull request. I have just taken the generated system types for now, I will go over them later and make sure they are all correct.

Also, I am absolutely not sure how the SDK works typescript-wise. I have just tried to add the types but I am not sure if I have done it in a good way, maybe you can take a look at that.

**Comment by moekify** [comment]
Hey, just checking in. Is this still something that is going to happen? Where did this PR end up after all? :)

**Comment by rijkvanzanten** [comment]
> Hey, just checking in. Is this still something that is going to happen? Where did this PR end up after all? :)

Definitely! I've just been really busy the last month or so, so haven't gotten around to reviewing most PRs yet 🙂

**Comment by NilsBaumgartner1994** [comment]
This is something we need !

**Comment by rijkvanzanten** [comment]
@CoenWarmer What are your thoughts on PascalCase vs underscores and singular vs plural? I'm currently feeling like we should return it however you have it configured in the data model to avoid unexpected behavior and make sure we prevent naming conflicts in the auto generated names, though that might result in non-conventional typescript types (like `type articles = { id: string }` rather than `type Article = { id: string }`

File: api/src/services/specifications.ts [hunk 4]
```diff
@@ -579,3 +582,120 @@ class GraphQLSpecsService implements SpecificationSubService {
...
+					};
+				})
+			)
+		).filter((collection) => collection.fields.length > 0);
+	}
+
+	async generateCollectionType(collection: { name: string; pascalName: string; fields: Field[] }) {
+		let collectionType = `export type ${collection.pascalName} = {\n`;
+		collectionType += Object.values(collection.fields)
+			.map((field) => `  ${field.field}${field.schema?.is_nullable ? '?' : ''}: ${this.getType(field.type)};`)
```
Comment (fehnomenal) [inline api/src/services/specifications.ts:639]: This should take into account relations to be usable with the sdk's `fields` option. Right now the types barf at something like `fields: ['image.id']`.

In my local version of this I use this:
image

**Review by fehnomenal — CHANGES_REQUESTED** [review]
I think the generated types should have the same case as in the app and the database to prevent surprises. The `CustomDirectusTypes` can be an exception. This way the "external" types are more clearly separated from your own types.

File: api/src/services/specifications.ts (hunk 4 above)
Comment (fehnomenal) [reply api/src/services/specifications.ts:639]: Well, in the meantime I adjusted my code after reading https://github.com/directus/sdk/discussions/34#discussioncomment-3359536 and to include m2m relations:

```ts
function getRawType(directusType: string): string {
  if (['integer', 'bigInteger', 'float', 'decimal'].includes(directusType)) return 'number';
  if (['boolean'].includes(directusType)) return 'boolean';
  if (['json'].includes(directusType)) return 'any[] | Record<string, any>';
  if (['csv'].includes(directusType)) return 'any[]';

  return 'string';
}

function getType(collection: string, field: Field, relations: Relation[]): string {
  const joinRelation = relations.find(
    (rel) => rel.related_collection === collection && rel.meta?.one_field === field.field,
  );
  if (joinRelation) {
    return `${getRawType(field.type)}[] | ${joinRelation.collection}[]`;
  }

  const relation = relations.find((rel) => rel.collection === collection && rel.field === field.field);
  if (relation?.related_collection) {
    return `${getRawType(field.type)} | ${relation.related_collection}`;
  }

  return getRawType(field.type);
}
```

My next step is to exclude types for relations the user has no access to.

**Comment by rijkvanzanten** [comment]
> I think the generated types should have the same case as in the app and the database to prevent surprises.

I have to agree here. Even though it's less common to see things like `blog_articles` in TypeScript, I'd rather keep it consistent and prevent naming collisions / unexpected quirks there (same as how we're handling things in GraphQL basically) 👍🏻

**Comment by marekbrze** [comment]
I hope this functionality will be added to the main project. I would love to use Directus instead of prisma in the T3 stack.

**Comment by jacobrienstra** [comment]
I recently used this in a fork of directus and came across a case that I wasn't sure how to deal with, and wanted to flag it here;

Obviously this needs to be updated a bit bc the sdk is no longer in this repo, but it does implement the system types and merges user provided types for system collections with those system types. 

The problem arises when the user has added custom fields to system collections, and _also_ references those system collections in relational fields. Because then in the types generated here, the user can only reference the system collection type that only includes custom fields, not the _merged_ type (that the SDK will create) that includes system and custom fields.

Ex:

```ts
type DirectusFiles = {
    custom_field: string;
}

type Articles = {
   featured_image: string | DirectusFiles // this only refers to the above DirectusFiles type....
}

export type DirectusCustomTypes = {
   directus_files: DirectusFiles;
   articles: Articles;
}
```

idk how to solve this really.... if the naming conventions are usual, could this reference `FileItem` (the merged type in the SDK)? It wouldn't compile on its own, but...

or I guess could you store the system types in shared (as is done here), the sdk could get those from shared, and if the user adds custom fields to system collections, this typing would include the system **and** custom fields. Then when it's intersected with the SDK types, it shouldn't be a problem bc the system fields will be exactly the same and therefore deduped....

**Pull request closed without merging**